test/gtest/ucp/test_ucp_proto_select.cc

```diff
@@ -0,0 +1,447 @@
+/**
+ * Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2024. ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "ucp_test.h"
+
+extern "C" {
+#include <ucp/core/ucp_ep.inl>
+#include <uct/base/uct_component.h>
+#include <uct/base/uct_iface.h>
+#include <uct/base/uct_md.h>
+#include <ucp/proto/proto_select.inl>
+#include <ucs/sys/topo/base/topo.h>
+}
+
+struct mock_t {
```

Comment:
If I understand correctly both `mock_t` and `mock_component` entities are used for two purposes where first is storage of mock and origin pointer or component and operations under them and another are their static interfaces which are supposed to store some global containers with `mock_t` or `mock_component` and making some operations with them.

So I want to propose to split these to API to two different entities: make `mock_t` and `mock_component` APIs as minimalistic as possible and move all the containers and operations with outside of the classes (maybe create separate workspace for them). To incapsulate map containers we can use unnamed namespaces.

Reply:
I refactored the code taking into account your feedback (also from our discussion). Key changes:
1. Split mock_t into outer mock { and inner struct mock_func }
2. Removed static global map - in fact we can instantiate mock per test, no need in global instance
3. Moved mock into test_helpers
4. Split mock_component into 2 entities:
- mock_component_test - This is the base class for component mocking, containing mock instance + implements global singleton map(), and setup_mock functions
- mock_component - this is API and infra for individual component mocking

Hope that helps